### ISSUE TITLE
[REVIEW]: CircuitGraph: A Python package for Boolean circuits

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2646]
**Submitting author:** @jpsety (<a href="http://orcid.org/0000-0002-3775-8638">Joseph Sweeney</a>)
**Repository:** <a href="https://github.com/circuitgraph/circuitgraph" target ="_blank">https://github.com/circuitgraph/circuitgraph</a>
**Version:** v0.0.3
**Editor:** @danielskatz
**Reviewer:** @skadio, @prw99r, @r-a-hoggarth
**Archive:** <a href="https://doi.org/10.5281/zenodo.4302741" target="_blank">10.5281/zenodo.4302741</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/33ed03caf258309f794875e9f4d1242e/status.svg)](https://joss.theoj.org/papers/33ed03caf258309f794875e9f4d1242e)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/33ed03caf258309f794875e9f4d1242e"><img src="https://joss.theoj.org/papers/33ed03caf258309f794875e9f4d1242e/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/33ed03caf258309f794875e9f4d1242e/status.svg)](https://joss.theoj.org/papers/33ed03caf258309f794875e9f4d1242e)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/circuitgraph/circuitgraph" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@skadio & @prw99r & @r-a-hoggarth, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/circuitgraph/circuitgraph">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jpsety) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/circuitgraph/circuitgraph">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jpsety) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/circuitgraph/circuitgraph">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jpsety) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @skadio, @prw99r, @r-a-hoggarth it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-94144-8_26 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02646/joss.02646/10.21105.joss.02646.pdf)

**Comment by danielskatz** [comment]
👋 @skadio, @prw99r, @r-a-hoggarth - Thanks again for agreeing to review.  Please read the comments above before starting.

This is the review thread for the paper. All of our communications will happen here from now on.

All reviewers have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention `openjournals/joss-reviews#2646` so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if any of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@danielskatz) if you have any questions/concerns.

**Comment by danielskatz** [comment]
👋 @skadio, @prw99r, @r-a-hoggarth - I just wanted to check on how things are going for you on your reviews. Let me know if I can do anything to help.

**Comment by danielskatz** [comment]
👋 @skadio, @prw99r, @r-a-hoggarth - Just checking in on this again...

**Comment by prw99r** [comment]
Just getting into the JOSS way of doing things - learning as I go...

**Comment by prw99r** [comment]
@whedon How to I edit my review checklist?

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by prw99r** [comment]
@danielskatz How do I edit my review checklist?

**Comment by danielskatz** [comment]
@whedon re-invite @prw99r as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@prw99r please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by r-a-hoggarth** [comment]
@whedon re-invite @r-a-hoggarth as reviewer

**Comment by whedon** [comment]
I'm sorry @r-a-hoggarth, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by danielskatz** [comment]
@whedon re-invite @r-a-hoggarth as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@r-a-hoggarth please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
It's good to see @r-a-hoggarth making progress, but I now wonder about @prw99r (no boxes checked) and @skadio (no changes recently).  How are things going?

**Comment by prw99r** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by prw99r** [comment]
Hi @danielskatz - how do I check off items on the checklist? I'm probably being really dumb, but how the heck do I do that? Thanks, Peter

**Comment by prw99r** [comment]
Hi Daniel, I've actually made some progress on looking at the code
(download the repo etc, having a good look at the code etc) - if I can
access the checklist (or there are some commands to do it) then I can
*look* like I'm actually doing something!
Cheers
Peter

**Comment by danielskatz** [comment]
@whedon re-invite @prw99r as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@prw99r please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
👋 @prw99r - you have to accept the invitation. See the comment above this one.

**Comment by prw99r** [comment]
it says the invitation  has expired?

**Comment by danielskatz** [comment]
@whedon re-invite @prw99r as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@prw99r please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
It must have been the one from the previous time expired between when I sent it and now - the new one should be available for 10-11 days, I think...  Please try again

**Comment by prw99r** [comment]
Got it - it works, I'm in!

**Comment by danielskatz** [comment]
👋 @skadio, @prw99r, @r-a-hoggarth - Can I again check in with you to see how things are going?  I see a bunch of boxes checked, but some not checked.  For each of you, are the unchecked boxes thing that you just haven't been able to check on yet, or are they items where there is work needed by the submitter, or discussion needed?  If so, please say what these blocking issues are.

Otherwise, can you tell me when you think you will be done with your reviews?

**Comment by danielskatz** [comment]
👋 @skadio, @prw99r, @r-a-hoggarth - ping... ⬆️

**Comment by danielskatz** [comment]
@r-a-hoggarth - it looks like you've checked off everything - can I take that to mean you are satisfied with the submission?

**Comment by skadio** [comment]
A quick update: I looked at the submissions but haven't completed the list yet, hoping to get back to it this weekend.

**Comment by prw99r** [comment]
I have some time scheduled this week to check things out some more.

**Comment by prw99r** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
> A quick update: I looked at the submissions but haven't completed the list yet, hoping to get back to it this weekend.

@skadio - any update?  (I know weekends aren't great for things like this, but just checking...)

**Comment by danielskatz** [comment]
👋 @prw99r & @skadio - any updates from you?  This review has been going about 2 months, and it would be really good to finish it soon.

I see a bunch of boxes checked, but some not checked. For each of you, are the unchecked boxes thing that you just haven't been able to check on yet, or are they items where there is work needed by the submitter, or discussion needed? If so, please say what these blocking issues are.

Otherwise, can you tell me when you think you will be done with your reviews?

**Comment by prw99r** [comment]
I've raised a couple of issues with the authors - I hope I'm not being too
picky, but trying to maintain a high standard. I'm not satisfied yet.
Peter

**Comment by danielskatz** [comment]
Thanks @prw99r - if you can put https://github.com/openjournals/joss-reviews/issues/2646 in a comment in those issue, then they become visible here and I can see if they are open or closed, which helps me track how this review is going.

**Comment by prw99r** [comment]
Daniel,
I've had a think, and I don't think I can approve this for JOSS publication
in its current form.
I took another look at the criteria for publication and guidelines and I
just dont think it satisfies the criteria laid down - it is an OK bit of
coding, as far as it goes, but a major scholarly contribution - no.

I'm not happy about the statement of need, references to prior work, and
also the level of contribution - I just don't see it.

Some useful packages - sure, but major contribution - nope.

It is not about raising specific issues either, that misses the point. it
is the fundamental contribution being asserted - I am not convinced it is
particularly interesting or useful other than to a very limited audience -
hence my comments about the statement of need.

Hope that helps,
Peter

**Comment by danielskatz** [comment]
Thanks @prw99r - I don't mean to minimize your concerns; I'm just trying to see where you have raised these issues and what the discussion has been on them.

**Comment by danielskatz** [comment]
Additionally, I wonder if @r-a-hoggarth has any thoughts on these issues.

**Comment by skadio** [comment]
@danielskatz  I completed this round of my review. My checklist is updated accordingly and some comments are added below for our authors @jpsety

I hear and share some of the valid points raised by @prw99r. At the moment, I am a bit on the fence with this submission. 

Overall, my understanding is that (please correct me if I am wrong) the library provides some I/O related capabilities to read/write circuit instances, an API to create/transform/interact with circuits, ability to do SAT queries. 

Under the hood, the bulk of the work is done by the `networkx` library to hold on to that structure and the `pysat` library to answer SAT questions. The standardization/synthesis/analysis are nice contributions but I am not yet convinced these grant an acceptance at JOSS. 

On the positive side, the package also ships with some circuit instances which others might find useful (assuming these are interesting instances for the community and are not available in standard formats elsewhere) and the library has the potential to "standardize" circuit operations to some extent. In that sense, one of the major contributions this library can have is "reproducibility" and sharing some common circuit formats --which can accelerate research. 

That said, as the authors rightly noted in their submission as well, the library is geared toward benefit the originating lab in the first place. I am not in a good position to fairly judge the applicability/interest from the rest of the community for this tool and/or whether these are gaps are not at all supported by other existing work in this field. 

=== 

- Comment about the codebase: while it seems okay overall but there are some dangling imports, typos, inconsistencies with PEP8 (or just linters in general). The source code can benefit from a quick clean up. 

- Codecov and coverage are listed as requirements, but I am not sure if there are necessary strict requirements. The setup.py does not list them either as install requirements so please consider simplifying the installation (or add them as required in setup.py) 

- Please explain what's `fanin` and `fanout` for the general reader. It is used in the Composition example without an explanation. 

- I couldn't find a contributor policy? 

- Have you considered load/save capabilities of Circuit objects? If researchers would like to share or persist the state of their circuit objects if saving/load to bytecode (say pickle files) conforms the integrity. This might be good to check/add. 

- @prw99r, I jsut realized that I might have mistakenly checked on/off some of the bullets from your checklist as I was scrolling up and down on the page. I hope I succeeded in retracted my edits but it might worth a quick look.

**Comment by danielskatz** [comment]
👋 @jpsety - Can I get your response to the comments from @prw99r & @skadio?

**Comment by danielskatz** [comment]
👋 @jpsety - Can I get your response to the comments from @prw99r & @skadio?

**Comment by jpsety** [comment]
@skadio, Thanks for the feedback, I've addressed most of your comments, but have some further questions/comments.

> Comment about the codebase: while it seems okay overall but there are some dangling imports, typos, inconsistencies with PEP8 (or just linters in general). The source code can benefit from a quick clean up.

I removed the unused imports/some variables I found. Ran the code through the Black linter as specified in  our contribution policy. Fixed typos.

> Codecov and coverage are listed as requirements, but I am not sure if there are necessary strict requirements. The setup.py does not list them either as install requirements so please consider simplifying the installation (or add them as required in setup.py)

Removed these from the requirements.txt, but this has broken our CI testing. Will find a fix. 

> Please explain what's fanin and fanout for the general reader. It is used in the Composition example without an explanation.

Added

> I couldn't find a contributor policy?

We have guidelines on the readme of the repo. 

> Have you considered load/save capabilities of Circuit objects? If researchers would like to share or persist the state of their circuit objects if saving/load to bytecode (say pickle files) conforms the integrity. This might be good to check/add.

I opened an issue for testing the pickle interface. This is certainly a convenient addition. I believe it should work already, but I'll add some unit tests to be sure. 

> Under the hood, the bulk of the work is done by the networkx library to hold on to that structure and the pysat library to answer SAT questions
I think this characterization misses the large amount of work in converting between formats, adding interfaces to generate circuits, transforms that enable SAT calls, etc. 

As per the JOSS guidelines:
> Your software should be a significant contribution to the available open source software that either enables some new research challenges to be addressed or makes addressing research challenges significantly better (e.g., faster, easier, simpler).
This library makes working with Boolean circuits significantly easier. In only a few lines of code, a user can manipulate circuits in ways that otherwise would take several languages, tools, etc. 
See the comment here: [comment](https://github.com/circuitgraph/circuitgraph/issues/18#issuecomment-724761050)

**Comment by jpsety** [comment]
@prw99r, I appreciate your feedback. You bring up good points in what can be added. I don't currently have the bandwidth to further edit this paper. As I understand it, the idea of the paper is to give a concise summary of the purpose of the library and show examples of the functionality. I have done that using other accepted projects as examples. 
I am confident that the targeted users of this library will be able to use it as is.

**Comment by danielskatz** [comment]
👋 @prw99r & @skadio - what are your thoughts on @jpsety's responses?

**Comment by skadio** [comment]
@danielskatz  

First of all, I would like to thank @jpsety for the updates and improvements. Most of these comments were not immediate blockers for me but I shared with you in a positive spirit to improve the work for the next/intended users beyond the lead developer. It's nice to see that some are incorporated and others can be added as well. 

Without going into a discussion of assessing the exact amount of work involved here, the author highlights an important aspect that I also agree with: "_the large amount of work in converting between formats, adding interfaces to generate circuits, transforms that enable SAT calls_". 

I consider this as a valuable contribution since it can help bring consistency across different research applications and for "reproducibility" purposes. In light of that, I don't see an immediate reason why should the research community be barred from learning about this tool and benefit from its functionality. Hence, I would be happy to accept this submission with minor edits from the list above, if any remaining. 

Best,
Serdar

**Comment by prw99r** [comment]
I agree with the highlighted contribution that @skadio notes above - I
think that @jpsety has gone partially down the road of dealing with my
concerns and added some useful clarifications. The purpose of our comments
has always been to provide positive criticism to improve the software and
the paper. It can be tough to take sometimes, especially if there is a lot
of time and effort already invested in the work, but it is worth taking the
time to do it right.

One nice aspect of this approach (JOSS) is that the issues are there for
all to see, and therefore it is left to the reader to make their own mind
up as to whether the paper (and ultimately the software) will do what they
need it to do, and if the paper is unclear, that they can ask the authors
to clarify further. The last thing I want to do, personally, is get in the
way of others using the library. In many respects that will be a/ the proof
that it is any good and b/ provide useful wider feedback.

I stand by my review comments, JOSS is a Journal after all, albeit in a
different format, and I don't see anything wrong with being rigorous and
demanding, However I do not have a problem with the library being released
for use by the design community, if that is the consensus, and therefore I
am happy to accept the submission now.

Peter

**Comment by danielskatz** [comment]
Thanks @skadio and @prw99r - Can I ask that check off any remaining items in your checklists in this case?  If there's anything that you feel you can't check off, let's talk about it here so that it's clear what is needed for acceptance (vs future enhancements that would be good but are not needed for acceptance)

Also, please feel free to open any needed issues in [the source repo](https://github.com/circuitgraph/circuitgraph) as that will be more of a record of potential future enhancement than this review issue, which will close (but remain available) after the review finishes.

**Comment by skadio** [comment]
Revised the Checklist, no need for opening source repo issues from my side.

**Comment by danielskatz** [comment]
👋 @prw99r - Can I ask that check off any remaining items in your checklist in [this case](https://github.com/openjournals/joss-reviews/issues/2646#issuecomment-732321669)? If there's anything that you feel you can't check off, let's talk about it here so that it's clear what is needed for acceptance (vs future enhancements that would be good but are not needed for acceptance)

Also, please feel free to open any needed issues in [the source repo](https://github.com/circuitgraph/circuitgraph) as that will be more of a record of potential future enhancement than this review issue, which will close (but remain available) after the review finishes.

**Comment by prw99r** [comment]
@danielskatz I have checked off my list now - my comments were for the paper more than the repo, so we are good to go.

**Comment by danielskatz** [comment]
Ok!  Thanks very much to everyone (@jpsety, @skadio, @prw99r, @r-a-hoggarth) for working through this process.  I think it's improved the software, and led to a good JOSS paper.

**Comment by danielskatz** [comment]
👋 @jpsety - At this point could you:
- [ ] Proofread the paper, and make sure you are happy with it (which I will also do after this part of the process)
- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by jpsety** [comment]
I've proofread the paper and followed the above instructions. 

Tagged release:
https://github.com/circuitgraph/circuitgraph/releases/tag/v0.0.3

DOI:
10.5281/zenodo.4302741

**Comment by danielskatz** [comment]
@whedon set v0.0.3 as version

**Comment by whedon** [comment]
OK. v0.0.3 is the version.

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.4302741 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4302741" target="_blank">10.5281/zenodo.4302741</a> is the archive.

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-94144-8_26 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
👋 @jpsety - some small final changes are needed in the paper and bib - please merge https://github.com/circuitgraph/circuitgraph/pull/22 or let me know what you disagree with.

**Comment by jpsety** [comment]
Merged the changes.
Thanks

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
sorry - one more change which I thought was in the PR but apparently wasn't - https://github.com/circuitgraph/circuitgraph/pull/23

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@jpsety - sorry, I missed a closing paren in the SciPy2008 bib entry - can you add it and then let me know - this really will be the last change...

**Comment by jpsety** [comment]
All set

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02646/joss.02646/10.21105.joss.02646.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-94144-8_26 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1963

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1963, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1334962699692892160) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1964
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02646](https://doi.org/10.21105/joss.02646)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by danielskatz** [comment]
Congratulations to @jpsety (Joseph Sweeney) and co-authors!!

And thanks very much to reviewers @skadio, @prw99r, @r-a-hoggarth!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02646/status.svg)](https://doi.org/10.21105/joss.02646)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02646">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02646/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02646/status.svg
   :target: https://doi.org/10.21105/joss.02646
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02646/status.svg)](https://doi.org/10.21105/joss.02646)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by skadio** [comment]
@danielskatz thank you as well for the organization and guiding us within a reasonable time-bounded review process. 

Congrats to @jpsety and the team!
